### PR TITLE
[TEST] Invoke tensor_filter with flexible input and output.

### PR DESCRIPTION
**Pull request by gichan-jang**
Testcase for flexible tensor filter is disabled and activate it after development is done.
In order to invoke tensor_filter with flexible input/output, you need to specify custom option.
 ex)  ``` ... ! tensor_filter framework=... invoke-dynamic=TRUE model=... ! ...```
 * Enable tests after implementation of the tensor_filter dynamic invoke

Self evaluation:

Build test: [ ]Passed [ ]Failed [*]Skipped
Run test: [ ]Passed [ ]Failed [*]Skipped

### CODE REVIEW

File: tests/nnstreamer_filter_custom/unittest_filter_custom.cc [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+  g_signal_connect (sink_handle, "new-data", (GCallback)new_data_cb, NULL);
+
+  filter_received = 0;
+  sink_received = 0;
+  EXPECT_EQ (setPipelineStateSync (gstpipe, GST_STATE_PLAYING, UNITTEST_STATECHANGE_TIMEOUT), 0);
+
+  EXPECT_TRUE (wait_pipeline_process_buffers (&sink_received, 9, TEST_TIMEOUT_LIMIT));
+
+  /** cleanup registered custom_easy filter */
+  ret = NNS_custom_easy_unregister ("safe_memcpy_10x10");
```
Comment (songgot) [inline tests/nnstreamer_filter_custom/unittest_filter_custom.cc:110]: "flexbible_filter" is correct, right?

Comment (gichan-jang) [reply tests/nnstreamer_filter_custom/unittest_filter_custom.cc]: Updated. Thanks.

File: gst/nnstreamer/include/nnstreamer_plugin_api_filter.h [hunk 2]
```diff
@@ -161,6 +161,8 @@ typedef struct _GstTensorFilterProperties
 
   int latency; /**< The average latency over the recent 10 inferences in microseconds */
   int throughput; /**< The average throughput in the number of outputs per second */
+
+  int invoke_dynamic; /**< True for supporting flexible invoke. NNFW plugin have to allocate output memory. */
```
Comment (songgot) [inline gst/nnstreamer/include/nnstreamer_plugin_api_filter.h:165]: 'plugins have' or 'plugin has' ?

File: gst/nnstreamer/tensor_filter/tensor_filter.c [hunk 3]
```diff
@@ -666,6 +666,12 @@ gst_tensor_filter_transform (GstBaseTransform * trans,
   out_flexible =
       gst_tensor_pad_caps_is_flexible (GST_BASE_TRANSFORM_SRC_PAD (trans));
 
+  if (priv->prop.invoke_dynamic && !out_flexible) {
```
Comment (songgot) [inline gst/nnstreamer/tensor_filter/tensor_filter.c:669]: In the case of (priv->prop.invoke_dynamic && !in_flexible), is confirmation required?

File: gst/nnstreamer/include/nnstreamer_plugin_api_filter.h [hunk 4]
```diff
@@ -295,6 +297,18 @@ struct _GstTensorFilterFramework
...
+  /**< Mandatory callback. Invoke the given network model.
+   *
+   * @param[in/out] private_data A subplugin may save its internal private data here. The subplugin is responsible for alloc/free of this pointer.
+   * @param[in] info structure of input tensors info
+   * @param[out] info structure of output tensors info. The subplugin should fill this info.
+   * @param[in] input The array of input tensors. Allocated and filled by tensor_filter/main
+   * @param[out] output The array of output tensors. The subplugin should allocate the memory block for output tensor. (data in GstTensorMemory)
+   * @return 0 if OK. non-zero if error.
+   */
+
```
Comment (myungjoo) [inline gst/nnstreamer/include/nnstreamer_plugin_api_filter.h:311]: The API format looks like v0 format (**private_data). Do not touch v0 format. Update v1 format.
And do not add in the middle of the struct. It breaks ABI of old binaries.

**Review by myungjoo — CHANGES_REQUESTED** [review]
Need to reconsider the interface change.

**Review by myungjoo — APPROVED** [review]
LGTM

**Review by myungjoo — COMMENTED** [review]
And make sure that the gstcap becomes ```format=flexible``` if it should be flexible ( framework supports flexible && the model wants flexible)

Even if framework supports flexible, if the models doesn't need flexible, try to keep it "static" for higher efficiency.,

File: gst/nnstreamer/tensor_filter/tensor_filter.c (hunk 3 above)
Comment (gichan-jang) [reply gst/nnstreamer/tensor_filter/tensor_filter.c]: I will check your comment in next PR. Thanks.

File: gst/nnstreamer/include/nnstreamer_plugin_api_filter.h (hunk 4 above)
Comment (gichan-jang) [reply gst/nnstreamer/include/nnstreamer_plugin_api_filter.h]: I removed `const` from `const prop` in invoke of filter v1 to update in/out tensors info.
If the sub-plugin using filter v0 wants to use a dynamic filter later, it needs to be changed to filter v1.
Thanks.

**Review by songgot — APPROVED** [review]
LGTM

**Pull request closed without merging**